cfme/utils/auth/__init__.py

```diff
@@ -0,0 +1,178 @@
+        ValueError if the yaml type for given key doesn't match auth_type on fetched class
+    """
+    auth_prov_config = auth_prov_data[auth_prov_key].copy()
+    klass = auth_class_from_type(auth_prov_config.get('type'))
+    if auth_prov_config.get('type') != klass.auth_type:
+        raise ValueError('{} must have type "{}"'.format(klass.__name__, klass.auth_type))
+    return klass.from_config(auth_prov_config, auth_prov_key)
+
+
+@attr.s
+class BaseAuthProvider(object):
+    """Base class for authentication provider objects
+
+    Might not be needed
+    """
+    auth_type = None
+    view_class = None
+    key = attr.ib()
+
+    @classmethod
+    def from_config(cls, prov_config, prov_key):
+        """Returns an object using the passed yaml config
+        Sets defaults for yaml configured objects separate from attr.ib definitions
+        """
+        prov_config.update(credentials[prov_config.get('credentials')])
+        class_attrs = [att.name for att in cls.__attrs_attrs__]
+        class_params = {k: v for k, v in iteritems(prov_config) if k in class_attrs}
+        return cls(key=prov_key, **class_params)
+
+    def as_fill_value(self):
+        class_attrs = [att.name for att in self.__attrs_attrs__]
+        include_attrs = [getattr(self.__class__, name)
+                         for name in self.view_class.cls_widget_names()
+                         if name in class_attrs]
+        return attr.asdict(self, filter=attr.filters.include(*include_attrs))
+
+
+@attr.s
+class AmazonAuthProvider(BaseAuthProvider):
+    """AWS IAM auth provider"""
+    auth_type = 'amazon'
+    view_class = AmazonAuthenticationView
+
+    access_key = attr.ib()
+    secret_key = attr.ib()
+    get_groups = attr.ib(default=False)
+
+
+@attr.s
+class MIQAuthProvider(BaseAuthProvider):
+    """base class for miq auth providers (ldap/ldaps modes in UI)
+    Intended to be used for freeipa, AD, openldap and openldaps type providers
+    """
+    auth_type = None
+    host1 = attr.ib()
+    bind_password = attr.ib()  # Ordered to adhere to mandatory attrs sequence
+    host2 = attr.ib(default=None)
+    host3 = attr.ib(default=None)
+    port = attr.ib(default=LDAP_PORT)
+    user_type = attr.ib(default='principal',
+                        validator=lambda item, attribute, value: value in USER_TYPES.keys())
+    domain_prefix = attr.ib(default=None)
+    user_suffix = attr.ib(default=None)
+    base_dn = attr.ib(default=None)
+    bind_dn = attr.ib(default=None)
+    get_groups = attr.ib(default=False)
+    get_roles = attr.ib(default=False)
+    follow_referrals = attr.ib(default=False)
+
+    # attrs for SSL
+    domain_name = attr.ib(default=None)
+    cert_filename = attr.ib(default=None)
+    cert_filepath = attr.ib(default=None)
+    ipaddress = attr.ib(default=None)
+    ldap_conf = attr.ib(default=None)
+    sssd_conf = attr.ib(default=None)
+
+    # TODO as_external_value method
+
+
+@attr.s
+class OpenldapAuthProvider(MIQAuthProvider):
```

Comment:
I'd make LDAP/IPA upper case since those are accronims

Reply:
ah you're right, LDAP is capitalized in openLDAP's brand materials. :+1: